### PR TITLE
Port LedgerApiServer rate limiter interface from canton

### LINKED ISSUE #9991: Allow ledger implementations to trigger backpressure before DAMLe is invoked
Currently, a ledger implementation can trigger backpressure by making `WriteService#submitTransaction` return `Overloaded`. This has the drawback that ledger can abort a command only after DAMLe has been executed. The current backpressure mechanism cannot be used to shield the system from excessive use of DAMLe.

To avoid that, extend `ApiCommandSubmissionService` with a callback `isOverloaded` that allows the ledger to tell if it is overloaded. Invoke `isOverloaded` before running DAMLe.

### PR DESCRIPTION
**Pull request by oliverse-da**
Fixes #9991 

changelog_begin
changelog_end

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [X] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by tudor-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ApiServices.scala [hunk 1]
```diff
@@ -95,6 +97,8 @@ private[daml] object ApiServices {
       seedService: SeedService,
       managementServiceTimeout: Duration,
       enableSelfServiceErrorCodes: Boolean,
+      checkOverloaded: TelemetryContext => Option[state.SubmissionResult] =
+        _ => None, // Used for Canton rate-limiting
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/ApiServices.scala:101]: I would prefer not to default the value at two different levels. Instead I would provide the default in SandboxServer.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/ApiSubmissionServiceSpec.scala [hunk 2]
```diff
@@ -489,6 +489,7 @@ object ApiSubmissionServiceSpec {
       errorCodesVersionSwitcher = new ErrorCodesVersionSwitcher(
         enableSelfServiceErrorCodes = useSelfServiceErrorCodes
       ),
+      checkOverloaded = _ => None, // Used and tested in Canton
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/services/ApiSubmissionServiceSpec.scala:492]: Why not add a test here as well? I.e., `ApiSubmissionService` and how the 'overload checking' function is used is still defined in the Daml SDK so this feature should be tested here as well.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala [hunk 3]
```diff
@@ -221,13 +224,20 @@ private[apiserver] final class ApiSubmissionService private[services] (
...
+          result <- commandExecutor.execute(commands, submissionSeed, ledgerConfig)
+          transactionInfo <- handleCommandExecutionResult(result)
+          partyAllocationResults <- allocateMissingInformees(transactionInfo.transaction)
+          submissionResult <- submitTransaction(
+            transactionInfo,
+            partyAllocationResults,
+            ledgerConfig,
+          )
+        } yield submissionResult
+    }
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:240]:
```suggestion
    }

```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala [hunk 4]
```diff
@@ -221,13 +224,20 @@ private[apiserver] final class ApiSubmissionService private[services] (
...
   ): Future[state.SubmissionResult] =
-    for {
-      result <- commandExecutor.execute(commands, submissionSeed, ledgerConfig)
-      transactionInfo <- handleCommandExecutionResult(result)
-      partyAllocationResults <- allocateMissingInformees(transactionInfo.transaction)
-      submissionResult <- submitTransaction(transactionInfo, partyAllocationResults, ledgerConfig)
-    } yield submissionResult
-
+    checkOverloaded(telemetryContext) match {
+      case Some(submissionResult) => Future.successful(submissionResult)
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:228]: Nit: another alternative may be to use `.map/.getOrElse`.
```scala
    checkOverloaded(telemetryContext)
      .map(Future.successful)
      .getOrElse {
        ...
      }
```

**Pull request merged**